heron/uploaders/src/java/com/twitter/heron/uploader/docker/DockerUploader.java

```diff
@@ -0,0 +1,187 @@
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package com.twitter.heron.uploader.docker;
+
+import com.twitter.heron.spi.common.Config;
+import com.twitter.heron.spi.common.Context;
+import com.twitter.heron.spi.uploader.IUploader;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URI;
+import java.util.UUID;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * IUploader Implementation that creates a DockerUploader Image for the Topology
+ * <p>
+ * The created DockerUploader Image will be based on the given base image and include the topology definition
+ * and binary in a fixed location. The uploader optionally pushes the docker image to a remote repository.
+ * Downstream schedulers must take responsibility for leveraging this image, by either building an additional
+ * DockerUploader image based on it with the correct START command
+ * </p>
+ * <p>
+ * The topology definition will be written to "/home/[role]/[topology-name]"
+ * if no role is given "heron" is used
+ * The URI returned by uploadPackage is the fully qualified tag name it will be
+ * "[docker-repository]/[cluster]/[role]/[env]/[topology-name]:[UUID]"
+ * if any of the values are not specified then they are omitted.
+ * UUIDs are used for the version in order to ensure that previously cached images won't get used.
+ * Per docker requirements role, env, and topology-name are all snake cased with Capital letters
+ * made lowercase and proceeded by a - unless the first character. for instance TopologyName would
+ * become topology-name
+ * </p>
+ * <p>
+ * The following configuration parameters are allowed:
+ * <dl>
+ * <dt>heron.uploader.docker.base</dt>
+ * <dd>The base docker image to use in the FROM directive, must be specified</dd>
+ * <dt>heron.uploader.docker.repository</dt>
+ * <dd>The repository prefix to used with the image tag, may be omitted</dd>
+ * <dt>heron.uploader.docker.push</dt>
+ * <dd>Boolean value if the docker image should be pushed to the remote repository, defaults to false</dd>
+ * </dl>
+ * </p>
+ * <p>
+ * The Dockerfile generated will be as follows:
+ * <code>
+ * FROM [base-image]
+ * ADD [topology.tar]   /home/[role | "heron"]/[topology-name]
+ * </code>
+ * Assuming your topology was named "MyTopology" and based on the ubuntu:trusty image with the role
+ * "MyRole" and packed into a tarfile called "MyTopology-3424.tar" the result would be.
+ * <code>
+ * FROM ubuntu:trusty
```

Comment:
you mean scheduler will download the docker image for heron core and the docker image for topology? or the heron core is a tar image (like we have it today) and the docker image for the topology will know where to download the core when docker image is starting? Is that correct?